turbopack/crates/turbopack-core/src/source_map/mod.rs

```diff
@@ -29,6 +29,14 @@ pub trait GenerateSourceMap {
     /// Generates a usable source map, capable of both tracing and stringifying.
     fn generate_source_map(self: Vc<Self>) -> Vc<OptionSourceMap>;
 
+    /// In case the source map is generated from a transformed asset, this returns the original
+    /// source.
+    ///
+    /// This is called only if `generate_source_map` returns a value.
+    fn original_source(self: Vc<Self>) -> Vc<OptionSource> {
```

Comment:
I added this here instead of `Source` because this behavior is specific to `generate_source_map`.

Reply:
It might be better to return `Some((source, source_map))` from `generate_source_map`, though.